pkg/autoscaler/collector.go

```diff
@@ -143,18 +189,43 @@ func (c *MetricCollector) Delete(ctx context.Context, namespace, name string) er
 	return nil
 }
 
+// Record records a stat that's been generated outside of the metric collector.
+func (c *MetricCollector) Record(key string, stat Stat) {
+	c.collectionsMutex.RLock()
+	defer c.collectionsMutex.RUnlock()
+
+	if collection, exists := c.collections[key]; exists {
```

Comment:
This perhaps needs a write lock?

Reply:
Why would it? It only ready the collections map. The `record` innerts protect themselves and are not part of that lock.

Reply:
Too many `record` functions. I looked at this one `c.buckets.Record(*stat.Time, stat.PodName, stat.AverageConcurrentRequests-stat.AverageProxiedConcurrentRequests)`. There was no mutex. I guess the `buckets.Record` does it. But the collector also has a mutex, so may be make buckets non-thread safe and push all locking to collector?

Way too many layers :-)